File: JVM/src/main/kotlin/io/alliance/adn/semantic/Lexer.kt
package io.alliance.adn.semantic

import io.alliance.adn.Walkable

internal class Lexer(input: List<Char>) : Walkable<Char>(input) {
    private val builder = StringBuilder()

    internal fun lex(): List<Token> {
        val result = ArrayList<Token>()

        while (index < size) {
            result.add(lexToken())
        }

        result.add(TokenStructural(TokenKind.MISC_END_OF_FILE))
        return result
    }

    private fun lexToken(): Token {
        while (current.isWhitespace()) {
            consume
        }

        when (current) {
            '=' -> {
                index++
                return TokenStructural(TokenKind.TOKEN_SET)
            }
            ',' -> {
                index++
                return TokenStructural(TokenKind.TOKEN_COMMA)
            }
            ':' -> {
                index++
                return TokenStructural(TokenKind.TOKEN_COLON)
            }
            ';' -> {
                index++
                return TokenStructural(TokenKind.TOKEN_SEMICOLON)
            }
            '{' -> {
                index++
                return TokenStructural(TokenKind.TOKEN_BRACE_OPEN)
            }
            '}' -> {
                index++
                return TokenStructural(TokenKind.TOKEN_BRACE_CLOSE)
            }
            '[' -> {
                if (next == ']') {
                    index += 2
                    return TokenStructural(TokenKind.KEYWORD_ARRAY)
                }
                index++
                return TokenStructural(TokenKind.TOKEN_BRACKET_OPEN)
            }
            ']' -> {
                index++
                return TokenStructural(TokenKind.TOKEN_BRACKET_CLOSE)
            }
            '-', '+', '.' -> {
                builder.append(consume)
                return lexNumber()
            }
            '"' ->  {
                index++
                return lexString()
            }
        }

        if (current.isDigit()) {
            return lexNumber()
        }

        if (current.isLetter()) {
            return lexTextual()
        }

        return TokenStructural(TokenKind.MISC_BAD_TOKEN)
    }

    private fun lexNumber(): Token {
        var bad = false
        var dot = false

        while (current.isDigit() || current == '.') {
            if (current == '.') {
                if (dot) {
                    bad = true
                }
                dot = true
            }

            builder.append(consume)
        }

        if (bad) {
            builder.clear()
            return TokenStructural(TokenKind.MISC_BAD_TOKEN)
        }

        val text = builder.toString()
        builder.clear()

        if (dot) {
            return TokenLiteral(TokenKind.LITERAL_FLOAT, text)
        }

        return TokenLiteral(TokenKind.LITERAL_NUMBER, text)
    }

    private fun lexString(): TokenLiteral {
        while (current != '"') {
            if (current == '\\') {
                when (next) {
                    '\\' -> builder.append('\\')
                    '\t' -> builder.append('\t')
                    '\r' -> builder.append('\r')
                    '\n' -> builder.append('\n')
                    '"' -> builder.append('"')
                    else -> {
                        builder.append('\\')
                        continue
                    }
                }

                index += 2
                continue
            }

            builder.append(consume)
        }

        index++

        val text = builder.toString()
        builder.clear()

        return TokenLiteral(TokenKind.LITERAL_STRING, text)
    }

    private fun lexTextual(): Token {
        while (current.isLetterOrDigit() || current == '_') {
            builder.append(consume)
        }

        val text = builder.toString()
        builder.clear()

        return when (text) {
            "struct" -> TokenStructural(TokenKind.KEYWORD_STRUCT)
            "bool" -> TokenType(TokenKind.KEYWORD_BOOL)
            "i8" -> TokenType(TokenKind.KEYWORD_I8)
            "i16" -> TokenType(TokenKind.KEYWORD_I16)
            "i32" -> TokenType(TokenKind.KEYWORD_I32)
            "i64" -> TokenType(TokenKind.KEYWORD_I64)
            "f32" -> TokenType(TokenKind.KEYWORD_F32)
            "f64" -> TokenType(TokenKind.KEYWORD_F64)
            "str" -> TokenType(TokenKind.KEYWORD_STR)

            "true" -> TokenLiteral.TRUE
            "false" -> TokenLiteral.FALSE
            else -> TokenIdentifier(text)
        }
    }
}

File: Native/src/semantic/lexer.rs
use std::cell::Cell;

use crate::semantic::TokenKind;
use crate::semantic::TokenKind::{Identifier, KeywordArray, KeywordBool, KeywordF32, KeywordF64, KeywordI16, KeywordI32, KeywordI64, KeywordI8, KeywordStr, KeywordStruct, LiteralBoolean, LiteralFloat, LiteralNumber, LiteralString, MiscBadToken, MiscEndOfFile, TokenBraceClose, TokenBraceOpen, TokenBracketClose, TokenBracketOpen, TokenColon, TokenComma, TokenSemicolon, TokenSet};
use crate::Walkable;

pub(crate) struct Lexer {
    list: Vec<char>,
    index: Cell<usize>,
}

impl Lexer {
    pub(crate) fn lex(input: String) -> Vec<TokenKind> {
        let lexer = Lexer {
            list: input.chars().collect(),
            index: Cell::new(0),
        };

        let mut tokens: Vec<TokenKind> = Vec::new();
        let max_index = lexer.list.len();

        while lexer.index.get() < max_index {
            tokens.push(match lexer.consume() {
                '{' => TokenBraceOpen,
                '[' => {
                    if lexer.current() == &']' {
                        lexer.skip();
                        KeywordArray
                    } else {
                        TokenBracketOpen
                    }
                },
                ']' => TokenBracketClose,
                '}' => TokenBraceClose,
                '=' => TokenSet,
                ',' => TokenComma,
                ';' => TokenSemicolon,
                ':' => TokenColon,
                '.' => lexer.lex_numeric(true, false),
                '+' => lexer.lex_numeric(false, false),
                '-' => lexer.lex_numeric(false, true),
                '"' => lexer.lex_string(),
                first => lexer.lex_textual(*first)
            })
        }

        tokens.push(MiscEndOfFile);

        tokens
    }

    fn lex_numeric(&self, dot: bool, neg: bool) -> TokenKind {
        let mut num: Vec<char> = Vec::new();

        if neg {
            num.push('-');
        }

        let mut dot = dot;
        let mut bad = false;

        loop {
            let cur = self.consume();

            if !cur.is_numeric() && cur != &'.' {
                break;
            }

            if cur == &'.' {
                if dot {
                    bad = true;
                }

                dot = true;
            }

            num.push(*cur);
        }

        let string = String::from_iter(num.iter());

        if bad {
            return MiscBadToken(string);
        }

        if dot {
            return LiteralFloat(string.parse::<f64>().expect("Error parsing float!"));
        }

        LiteralNumber(string.parse::<i64>().expect("Error parsing integer!"))
    }

    /// - Opening and closing quotation marks are not included in the string.
    /// - Tab, carriage return, line feed, quotation mark and backslash may be escaped (with a backslash).
    /// - A backslash that does not escape one of the five characters will be ignored and NOT included in the string.
    fn lex_string(&self) -> TokenKind {
        let mut text: Vec<char> = Vec::new();

        loop {
            let current = self.consume();

            if current == &'\"' {
                break;
            }

            if current == &'\\' {
                match self.consume() {
                    '"' => text.push('"'),
                    't' => text.push('\t'),
                    'r' => text.push('\r'),
                    'n' => text.push('\n'),
                    '\\' => text.push('\\'),
                    char => text.push(*char)
                }
            }

            text.push(*current);
        }

        return LiteralString(String::from_iter(text.iter()));
    }

    fn lex_textual(&self, first: char) -> TokenKind {
        if !first.is_alphabetic() {
            return MiscBadToken(String::from(first));
        }

        let mut text: Vec<char> = vec![first];

        loop {
            let current = self.current();

            if !current.is_alphanumeric() && current != &'_' {
                break;
            }

            text.push(*current);
            self.skip();
        }

        let text = String::from_iter(text.iter());

        match &*text {
            "struct" => KeywordStruct,
            "bool" => KeywordBool,
            "i8" => KeywordI8,
            "i16" => KeywordI16,
            "i32" => KeywordI32,
            "i64" => KeywordI64,
            "f32" => KeywordF32,
            "f64" => KeywordF64,
            "str" => KeywordStr,

            "true" => LiteralBoolean(true),
            "false" => LiteralBoolean(false),

            _ => Identifier(text),
        }
    }
}

impl Walkable<char> for Lexer {
    #[inline]
    fn size(&self) -> usize {
        self.list.len()
    }

    #[inline]
    fn peek(&self, offset: usize) -> &char {
        self.list.get(self.index.get() + offset).expect("Index out of bounds!")
    }

    #[inline]
    fn skip(&self) {
        self.index.set(self.index.get() + 1);
    }
}